sort/merge_sort.go

```diff
@@ -0,0 +1,34 @@
+package sort
+
+func mergeSort(arr []int) []int {
+	if len(arr) <= 1 {
+		return arr
+	}
+	mid := len(arr) / 2
+	left := make([]int, mid)
+	right := make([]int, len(arr)-mid)
+
+	copy(left, arr[:mid])
+	copy(right, arr[mid:])
+
+	left = mergeSort(left)
+	right = mergeSort(right)
+
+	return merge(left, right)
+}
+
+func merge(left, right []int) []int {
+	ret := make([]int, 0)
+	for len(left) > 0 && len(right) > 0 {
```

Comment:
### Optimizations

The `merge` function currently appends remaining elements from either `left` or `right` after the main loop. This can be optimized by preallocating the result slice to the exact size needed. This avoids multiple reallocations and improves performance:

```go
ret := make([]int, 0, len(left) + len(right))
```